File: packages/logger/src/faro-engine.ts
import type { LogLevel, LogRecord, SpanRecord, TelemetrySink } from './types.js'

/**
 * Faro-backed engine. `@grafana/faro-web-sdk` + `@grafana/faro-web-tracing`
 * are **optional peerDependencies** — they are loaded dynamically and never
 * referenced in this module's public types. If the peers are absent the
 * factory resolves to `null` so the caller can fall back to console.
 *
 * For tests, a `transport` may be injected: an object with a `push(payload)`
 * method. This bypasses Faro entirely (no network, no peer required).
 */

/** Minimal structural shape of a Faro-ish transport. Not exported. */
interface FaroTransport {
  push(payload: { kind: 'log' | 'span'; record: LogRecord | SpanRecord }): void
}

export interface FaroEngineOptions {
  app: string
  endpoint: string
  /**
   * Test/override transport. When provided, the Faro peers are NOT loaded
   * and records are forwarded straight to `transport.push`.
   */
  transport?: FaroTransport
}

/** Map our levels onto Faro's log-level strings. */
const FARO_LEVEL: Record<LogLevel, string> = {
  debug: 'debug',
  info: 'info',
  warn: 'warn',
  error: 'error',
  fatal: 'error',
}

/**
 * Construct the Faro engine. Returns `null` when the optional peers are not
 * installed and no override transport was supplied — callers treat `null` as
 * "fall back to console".
 */
export async function createFaroSink(
  opts: FaroEngineOptions,
): Promise<TelemetrySink | null> {
  const transport = opts.transport ?? (await loadFaroTransport(opts))
  if (!transport) return null

  return {
    name: 'faro',

    log(record: LogRecord): void {
      transport.push({ kind: 'log', record })
    },

    span(record: SpanRecord): void {
      transport.push({ kind: 'span', record })
    },

    async flush(): Promise<void> {
      // Faro's own transports flush on their schedule / on beacon; the
      // manager drives page-exit beacon flushing. Nothing buffered here.
    },
  }
}

/**
 * Dynamically import the Faro peers and adapt them to {@link FaroTransport}.
 * Returns `null` if either peer is missing (optional peerDependency absent).
 */
async function loadFaroTransport(
  opts: FaroEngineOptions,
): Promise<FaroTransport | null> {
  try {
    // Indirected so bundlers keep these as runtime-optional dynamic imports.
    const sdkName = '@grafana/faro-web-sdk'
    const tracingName = '@grafana/faro-web-tracing'
    const sdk = (await import(/* @vite-ignore */ sdkName)) as {
      initializeFaro: (cfg: unknown) => unknown
      getWebInstrumentations: () => unknown[]
    }
    const tracing = (await import(/* @vite-ignore */ tracingName)) as {
      TracingInstrumentation: new () => unknown
    }

    const faro = sdk.initializeFaro({
      url: opts.endpoint,
      app: { name: opts.app },
      instrumentations: [
        ...sdk.getWebInstrumentations(),
        new tracing.TracingInstrumentation(),
      ],
    }) as {
      api: {
        pushLog: (msgs: unknown[], opts?: unknown) => void
        pushEvent: (name: string, attrs?: Record<string, unknown>) => void
      }
    }

    return {
      push({ kind, record }): void {
        if (kind === 'log') {
          const r = record as LogRecord
          faro.api.pushLog([r.message], {
            level: FARO_LEVEL[r.level],
            context: flatten(r.context),
          })
        } else {
          const r = record as SpanRecord
          faro.api.pushEvent(`span:${r.name}`, {
            durationMs: String(r.durationMs),
            status: r.status,
            ...flatten(r.context),
          })
        }
      },
    }
  } catch {
    // Peer not installed (optional) or init failed — caller falls back.
    return null
  }
}

/** Faro context attributes are flat string maps; coerce ours to match. */
function flatten(ctx: Record<string, unknown>): Record<string, string> {
  const out: Record<string, string> = {}
  for (const [k, v] of Object.entries(ctx)) {
    out[k] = typeof v === 'string' ? v : JSON.stringify(v)
  }
  return out
}

File: packages/flutter/lib/src/telemetry/faro_engine.dart
/// Faro-backed engine — a 1:1 port of `@refraction-ui/logger`
/// `faro-engine.ts`.
///
/// The vendor (Faro) is an **internal abstraction behind [TelemetrySink]** —
/// no vendor type appears in this module's public surface. The real
/// Faro-Flutter / OTel-Dart binding is a Wave 1 concern: when no override
/// [FaroEngineOptions.transport] is supplied this factory resolves to `null`
/// (peer absent), so the caller falls back to console — identical to the web
/// semantics where the optional peer may be missing. The on-wire envelope
/// (`{ kind, record }` pushed to a transport) is identical to web (OTLP).
library;

import 'types.dart';

/// Minimal structural shape of a Faro-ish transport. Not part of the public
/// vendor-neutral API beyond this engine's options. The payload envelope is
/// identical to the web engine.
abstract class FaroTransport {
  /// Receive a `{ kind, record }` envelope. [kind] is `'log'` or `'span'`;
  /// [record] is a [LogRecord] or [SpanRecord] respectively.
  void push({required String kind, required Object record});
}

/// Options for [createFaroSink].
class FaroEngineOptions {
  /// Creates [FaroEngineOptions].
  const FaroEngineOptions({
    required this.app,
    required this.endpoint,
    this.transport,
  });

  /// Logical app/service name.
  final String app;

  /// Consumer-supplied collector endpoint. Data goes to the consumer's
  /// endpoint, never refraction-ui's.
  final String endpoint;

  /// Test/override transport. When provided, the Faro vendor is NOT loaded
  /// and records are forwarded straight to [FaroTransport.push].
  final FaroTransport? transport;
}

/// Map our levels onto Faro's log-level strings (`fatal` -> `error`).
const Map<LogLevel, String> _faroLevel = <LogLevel, String>{
  LogLevel.debug: 'debug',
  LogLevel.info: 'info',
  LogLevel.warn: 'warn',
  LogLevel.error: 'error',
  LogLevel.fatal: 'error',
};

class _FaroSink implements TelemetrySink {
  _FaroSink(this._transport);

  final FaroTransport _transport;

  @override
  String get name => 'faro';

  @override
  void log(LogRecord record) {
    _transport.push(kind: 'log', record: record);
  }

  @override
  void span(SpanRecord record) {
    _transport.push(kind: 'span', record: record);
  }

  @override
  Future<void> flush() async {
    // The vendor's own transport flushes on its schedule / on background;
    // the manager drives app-exit flushing. Nothing buffered here.
  }
}

/// Construct the Faro engine. Returns `null` when no vendor binding is
/// available and no override [FaroEngineOptions.transport] was supplied —
/// callers treat `null` as "fall back to console". The real Faro-Flutter /
/// OTel-Dart binding lands in Wave 1 behind this same surface.
Future<TelemetrySink?> createFaroSink(FaroEngineOptions opts) async {
  final transport = opts.transport ?? await _loadFaroTransport(opts);
  if (transport == null) return null;
  return _FaroSink(transport);
}

/// Adapt the (Wave 1) Faro vendor to [FaroTransport]. Returns `null` while
/// no vendor binding is wired (Wave 0) — exactly mirroring the web behavior
/// where the optional peer is absent.
Future<FaroTransport?> _loadFaroTransport(FaroEngineOptions opts) async {
  // Wave 0: no vendor binding. Wave 1 plugs Faro-Flutter / OTel-Dart in here
  // without any change to this engine's public surface or the OTLP envelope.
  return null;
}

/// Faro context attributes are flat string maps; coerce ours to match.
/// Exposed for parity tests against the web engine's `flatten`.
Map<String, String> flattenContext(Map<String, Object?> ctx) {
  final out = <String, String>{};
  ctx.forEach((k, v) {
    out[k] = v is String ? v : _jsonish(v);
  });
  return out;
}

String _jsonish(Object? v) {
  if (v == null) return 'null';
  if (v is num || v is bool) return v.toString();
  if (v is String) return v;
  // Match JSON.stringify for collections.
  return _stringify(v);
}

String _stringify(Object? v) {
  if (v == null) return 'null';
  if (v is String) return '"$v"';
  if (v is num || v is bool) return v.toString();
  if (v is List) {
    return '[${v.map(_stringify).join(',')}]';
  }
  if (v is Map) {
    final entries = v.entries
        .map((e) => '"${e.key}":${_stringify(e.value)}')
        .join(',');
    return '{$entries}';
  }
  return '"$v"';
}

/// Maps a [LogLevel] to the Faro log-level string. Exposed for parity tests.
String faroLevelName(LogLevel level) => _faroLevel[level]!;
